modules/porous_flow/include/actions/PorousFlowActionBase.h

```diff
@@ -0,0 +1,195 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef POROUSFLOWACTIONBASE_H
+#define POROUSFLOWACTIONBASE_H
+
+#include "Action.h"
+#include "PorousFlowDependencies.h"
+
+class PorousFlowActionBase;
+
+template <>
+InputParameters validParams<PorousFlowActionBase>();
+
+/**
+ * Base class for PorousFlow actions.  This act() method simply
+ * defines the name of the PorousFlowDictator.  However, this
+ * class also contains a number of utility functions that may
+ * be used by derived classes
+ */
+class PorousFlowActionBase : public Action, public PorousFlowDependencies
+{
+public:
+  PorousFlowActionBase(const InputParameters & params);
+
+  virtual void act() override;
+
+protected:
+  /**
+   * List of Kernels, AuxKernels, Materials, etc, to be added.
+   * This list will be used to determine what Materials need
+   * to be added.
+   * Actions may add or remove things from this list
+   */
+  std::vector<std::string> _objects_to_add;
+
+  /// The name of the PorousFlowDictator object to be added
+  const std::string _dictator_name;
+
+  /// gravity
+  const RealVectorValue _gravity;
+
+  /// Name of the mass-fraction variables (if any)
+  const std::vector<VariableName> _mass_fraction_vars;
+
+  /// Number of mass-fraction variables
+  const unsigned _num_mass_fraction_vars;
+
+  /// Name of the temperature variable (if any)
+  const std::vector<VariableName> _temperature_var;
+
+  /// displacement NonlinearVariable names (if any)
+  const std::vector<NonlinearVariableName> _displacements;
+
+  /// number of displacement variables supplied
+  const unsigned _ndisp;
+
+  /// displacement Variable names
+  std::vector<VariableName> _coupled_displacements;
```

Comment:
You can store references to all these names instead of making copies. This won't cause any problems in your code but just saves a little space since the InputParameters object sticks around for the lifetime of your object.